markovflow/models/spatio_temporal_variational.py

```diff
@@ -270,12 +341,256 @@ def posterior(self) -> PosteriorProcess:
+                for l in tf.dynamic_partition(theta_linear, indices, num_partitions=num_partition)
+            ]
+        )
+        summed_lik_nat2 = tf.stack(
+            [
+                tf.reduce_sum(l, axis=0)
+                for l in tf.dynamic_partition(lik_nat2, indices, num_partitions=num_partition)
+            ]
+        )
+
+        # update
+        lr = self.learning_rate
+        new_nat1 = (1 - lr) * self.nat1 + lr * summed_theta_linear
+        new_nat2 = (1 - lr) * self.nat2 + lr * summed_lik_nat2
+
+        self.nat2.assign(new_nat2)
+        self.nat1.assign(new_nat1)
+
+    def loss(self, input_data: Tuple[tf.Tensor, tf.Tensor]) -> tf.Tensor:
+        """
+        Obtain a `Tensor` representing the loss, which can be used to train the model.
+
+        :param input_data: A tuple of time points and observations containing the data at which
+            to calculate the loss for training the model.
+        """
+        return -self.elbo(input_data)
+
+    def local_objective_and_gradients(
+        self, Fmu: tf.Tensor, Fvar: tf.Tensor, Y: tf.Tensor
     ) -> tf.Tensor:
         """
-        Compute the log density of the data at the new data points.
+        Returs the local_objective and its gradients wrt to the expectation parameters
+        :param Fmu: means μ [..., latent_dim]
+        :param Fvar: variances σ² [..., latent_dim]
+        :param Y: observations Y [..., observation_dim]
+        :return: local objective and gradient wrt [μ, σ² + μ²]
         """
-        X, Y = input_data
-        f_mean, f_var = self.space_time_predict_f(X)
-        return self.likelihood.predict_log_density(f_mean, f_var, Y)
+
+        with tf.GradientTape() as g:
+            g.watch([Fmu, Fvar])
+            local_obj = tf.reduce_sum(input_tensor=self.local_objective(Fmu, Fvar, Y))
+        grads = g.gradient(local_obj, [Fmu, Fvar])
+
+        # turn into gradient wrt μ, σ² + μ²
+        grads = gradient_transformation_mean_var_to_expectation([Fmu, Fvar], grads)
+
+        return local_obj, grads
+
+    def local_objective(self, Fmu: tf.Tensor, Fvar: tf.Tensor, Y: tf.Tensor) -> tf.Tensor:
+        """
+        local loss in CVI
+        :param Fmu: means [..., latent_dim]
+        :param Fvar: variances [..., latent_dim]
+        :param Y: observations [..., observation_dim]
+        :return: local objective [...]
+        """
+        return self._likelihood.variational_expectations(Fmu, Fvar, Y)
+
+    @property
+    def kernel(self) -> SDEKernel:
+        """
+        Return the kernel of the GP.
+        """
+        return self._kernel
+
+    @property
+    def inducing_time(self) -> tf.Tensor:
+        """
+        Return the temporal inducing inputs of the model.
+        """
+        return self._inducing_time
+
+    @property
+    def inducing_space(self) -> tf.Tensor:
+        """
+        Return the spatial inducing inputs of the model.
+        """
+        return self._inducing_space
```

Comment:
These getters are missing from `SpatioTemporalSparseVariational`. Since the returned variables are stored in the base class, should these getters be moved to the base class?